crates/kernel/src/tasks/scheduler.rs

```diff
@@ -1045,142 +1104,63 @@ impl Scheduler {
-        let forked_session = session.clone();
-        let task_id = self
-            .submit_fork_task(fork_request, forked_session)
-            .unwrap_or_else(|e| panic!("Could not fork task: {:?}", e));
-
-        let reply = reply;
-        if let Err(e) = reply.send(task_id) {
-            error!(task = task_id, error = ?e, "Could not send fork reply. Parent task gone?  Remove.");
-            to_remove.push(task_id);
-        }
-        to_remove
-    }
-
-    fn process_describe_request(
-        &self,
-        requesting_task_id: TaskId,
-        reply: oneshot::Sender<Vec<TaskDescription>>,
-    ) -> Vec<TaskId> {
-        let mut to_remove = vec![];
-
-        let reply = reply;
-
-        // Note these could be done in parallel and joined instead of single file, to avoid blocking
-        // the loop on one uncooperative thread, and could be done in a separate thread as well?
-        // The challenge being the borrow semantics of the 'tasks' list.
-        // And we should have a timeout here to boot.
-        // For now, just iterate blocking.
-        let mut tasks = Vec::new();
-        trace!(
-            task = requesting_task_id,
-            "Task requesting task descriptions"
-        );
-        let inner = self.inner.read().unwrap();
-        for (task_id, task) in inner.tasks.iter() {
-            // Tasks not in suspended state shouldn't be added.
-            if !task.suspended {
-                continue;
-            }
-            if *task_id != requesting_task_id {
-                trace!(
-                    requesting_task_id = requesting_task_id,
-                    other_task = task_id,
-                    "Requesting task description"
-                );
-                let (t_send, t_reply) = oneshot::channel();
-                let tcs = task.task_control_sender.clone();
-                if let Err(e) = tcs.send(TaskControlMsg::Describe(t_send)) {
-                    error!(?task_id, error = ?e,
-                            "Could not send describe request to task. Task being removed.");
-                    to_remove.push(task.task_id);
-                    continue;
-                }
-                let Ok(task_desc) = t_reply.recv() else {
-                    error!(?task_id, "Could not get task description");
-                    to_remove.push(task.task_id);
-                    continue;
-                };
-                trace!(
-                    requesting_task_id = requesting_task_id,
-                    other_task = task_id,
-                    "Got task description"
-                );
-                tasks.push(task_desc);
-            }
+    #[instrument(skip(self, control_sender, database))]
+    fn retry_task(
+        &mut self,
+        task: Task,
+        control_sender: &Sender<(TaskId, SchedulerControlMsg)>,
+        database: Arc<dyn Database>,
+    ) {
+        // Make sure the old thread is dead.
+        task.kill_switch.store(false, Ordering::SeqCst);
```

Comment:
I am confusion, how does setting the killswitch to *`false`* make sure the old thread is dead?

Reply:
oops, thanks for catching this!